library/src/main/java/bio/terra/pfb/PfbReader.java

```diff
@@ -0,0 +1,182 @@
+package bio.terra.pfb;
+
+import bio.terra.pfb.exceptions.InvalidPfbException;
+import java.io.*;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileReader;
+import org.apache.avro.file.DataFileStream;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.specific.SpecificDatumReader;
+
+public class PfbReader {
+
+  public String showSchema(String fileLocation) throws IOException {
+    Schema schema = getSchema(fileLocation);
+    List<Schema> pfbSchemaShow =
+        schema.getField("object").schema().getTypes().stream()
+            .filter(t -> !t.getName().equals("Metadata"))
+            .toList();
+    String show = pfbSchemaShow.stream().map(s -> s.toString()).toList().toString();
+    return convertEnum(show);
+  }
+
+  public String showNodes(String fileLocation) throws IOException {
+    Metadata metadata = deserializeAvroFile(fileLocation);
+    return metadata.getNodes().stream().map(n -> n.getName()).collect(Collectors.joining("\n"))
+        + "\n";
+  }
+
+  public String show(String fileLocation) throws IOException {
+    // TODO - figure out how to read the rest of the Entity objects into a list of single json
+    // objects
+    return "NOT IMPLEMENTED";
+  }
+
+  public String showMetadata(String fileLocation) throws IOException {
+    Metadata metadata = deserializeAvroFile(fileLocation);
+    return metadata.toString();
+  }
+
+  public Schema getSchema(String fileLocation) throws IOException {
+    boolean isUrl = isValidUrl(fileLocation);
+    if (isUrl) {
+      return readUrlPFBSchema(fileLocation);
+    }
+    return readFilePathPFBSchema(fileLocation);
+  }
+
+  Schema readFilePathPFBSchema(String fileLocation) throws IOException {
+    DatumReader<Entity> datumReader = new SpecificDatumReader<>(Entity.class);
+    DataFileReader<Entity> dataFileReader =
+        new DataFileReader<>(new File(fileLocation), datumReader);
+    return dataFileReader.getSchema();
+  }
+
+  public Metadata deserializeAvroFile(String fileLocation) throws IOException {
```

Comment:
Is an avro file just Metadata?  Would it make sense to name this something like 'getMetadata' since that's what it's aways doing, or is there never a case where you would deserialize and avro file and get something other than metadata?

Reply:
This method focuses on Metadata, so yes, good idea to rename it. 

PFB files are Avro files that follow a specific schema. The schema includes a little of "Entity" objects, where an entity can either be a "Metadata" object or a table row object.  I've tried to detail this setup in this diagram and there are more details in the [pyPFB library](https://github.com/uc-cdis/pypfb/blob/master/docs/index.md#introduction). 
![PFB](https://github.com/DataBiosphere/java-pfb/assets/13254229/40bc95b0-7f23-4bd6-870e-9dcb34aa89c5)